### PR TITLE
Telemetry hook

### PR DESCRIPTION
**Pull request by raboof**
WIP: only scaladsl so far

### CODE REVIEW

File: runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+          NoOpTelemetry
+      }
+    }
+  }
+}
+
+@InternalStableApi
+trait TelemetrySpi {
+  @nowarn
+  def onRequest(prefix: String, method: String, request: HttpRequest): Unit = {}
```
Comment (johanandren) [inline runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala:55]: May need to be able to modify request and return a new one to pass context. You'll see later.

**Review by johanandren — COMMENTED** [review]
Looks good so far.

File: runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala (hunk 1 above)
Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala:55]: :+1: I suspect not for *this* feature, but indeed would be a nice generalization

File: runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala [hunk 2]
```diff
@@ -52,7 +52,7 @@ private[internal] object TelemetrySpi {
 @InternalStableApi
 trait TelemetrySpi {
   @nowarn
-  def onRequest(prefix: String, method: String, request: HttpRequest): Unit = {}
+  def onRequest[T <: HttpRequest](prefix: String, method: String, request: T): T = request
```
Comment (johanandren) [inline runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala:55]: Why was the type bound needed, something on the "other side"?

File: interop-tests/src/test/scala/akka/grpc/scaladsl/TelemetrySpec.scala [hunk 3]
```diff
@@ -37,14 +40,16 @@ class TelemetrySpec
...
       prefix should be(GreeterService.name)
       method should be("SayHello")
-      request.entity.contentType should be(GrpcProtocolNative.contentType)
+      request.entity.getContentType should be(GrpcProtocolNative.contentType)
     }
   }
 }
 
 class CollectingTelemetrySpi extends TelemetrySpi {
   var requests: List[(String, String, HttpRequest)] = Nil
```
Comment (johanandren) [inline interop-tests/src/test/scala/akka/grpc/scaladsl/TelemetrySpec.scala:49]: @volatile perhaps if it should be safe to observe from another thread

Comment (raboof) [reply interop-tests/src/test/scala/akka/grpc/scaladsl/TelemetrySpec.scala:49]: this is just the spec, but yeah, good to add

File: runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala (hunk 2 above)
Comment (raboof) [reply runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala:55]: yeah, so we can keep using the scala HttpRequest as a scala HttpRequest after going through the SPI

File: runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  ExtensionIdProvider
+}
+import akka.annotation.{ InternalApi, InternalStableApi }
+import akka.http.javadsl.model.HttpRequest
+
+import scala.annotation.nowarn
+
+class TelemetryExtensionImpl(val spi: TelemetrySpi) extends Extension
+
+object TelemetryExtension extends ExtensionId[TelemetryExtensionImpl] with ExtensionIdProvider {
```
Comment (ignasi35) [inline runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala:22]:
```
/** INTERNAL API */
@InternalStableApi
private[grpc] ...
```

(both class and object)

**Review by ignasi35 — APPROVED** [review]

**Review by johanandren — APPROVED** [review]

**Pull request merged**